src/Common/Core/Impl/Services/ServiceManager.cs

```diff
@@ -28,22 +28,33 @@ public class ServiceManager : IServiceManager {
         public virtual IServiceManager AddService<T>(T service, Type type = null) where T : class {
             _disposeToken.ThrowIfDisposed();
 
-            type = type ?? typeof(T);
+            type = type ?? service.GetType();
```

Comment:
This change creates several problems:
1. It pretty much removes all the benefits of using the dictionary, cause most of the times service is requested by interface, not by type itself, so to find required instance `ServiceManager` will have to iterate through collection all the time.
2. Inconsistency. For `class C : I1, I2` `ServiceManager.AddService<I1>(new C())` and `ServiceManager.AddService<I1>(() => new C1())` will create different results.

It looks like this method has to be changed to:
```cs
IServiceManager AddService(object instance, Type type = null)
```
which will match `void RemoveService(object service)`. And then we may have an extension:
```cs
IServiceManager AddService<T>(this IServiceManager serviceManager, T instance) where T : class => serviceManager.AddService(instance, typeof(T));
```